src/erc-721/ERC721TLMutable.sol

```diff
@@ -0,0 +1,458 @@
+        _counter += numTokens;
+        _batchMints.push(BatchMint(recipient, start, end, baseUri));
+
+        _increaseBalance(recipient, numTokens); // this function adds the number of tokens to the recipient address
+
+        for (uint256 id = start; id < end + 1; ++id) {
+            emit Transfer(address(0), recipient, id);
+        }
+    }
+
+    /// @inheritdoc IERC721TL
+    function airdrop(address[] calldata addresses, string calldata baseUri) external onlyRoleOrOwner(ADMIN_ROLE) {
+        if (bytes(baseUri).length == 0) revert EmptyTokenURI();
+        if (addresses.length < 2) revert AirdropTooFewAddresses();
+
+        uint256 start = _counter + 1;
+        uint256 end = start + addresses.length - 1;
+        _counter += addresses.length;
+        _batchMints.push(BatchMint(address(0), start, end, baseUri));
+        for (uint256 i = 0; i < addresses.length; i++) {
+            _mint(addresses[i], start + i);
+        }
+    }
+
+    /// @inheritdoc IERC721TL
+    function externalMint(address recipient, string calldata uri) external onlyRole(APPROVED_MINT_CONTRACT) {
+        if (bytes(uri).length == 0) revert EmptyTokenURI();
+        _counter++;
+        _tokenUris[_counter] = uri;
+        _mint(recipient, _counter);
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                Burn Functions
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IERC721TL
+    function burn(uint256 tokenId) external {
+        address tokenOwner = ownerOf(tokenId);
+        if (!_isAuthorized(tokenOwner, msg.sender, tokenId)) revert CallerNotApprovedOrOwner();
+        _burn(tokenId);
+        _burned[tokenId] = true;
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                Royalty Functions
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc ICreatorBase
+    function setDefaultRoyalty(address newRecipient, uint256 newPercentage) external onlyRoleOrOwner(ADMIN_ROLE) {
+        _setDefaultRoyaltyInfo(newRecipient, newPercentage);
+    }
+
+    /// @inheritdoc ICreatorBase
+    function setTokenRoyalty(uint256 tokenId, address newRecipient, uint256 newPercentage)
+        external
+        onlyRoleOrOwner(ADMIN_ROLE)
+    {
+        _overrideTokenRoyaltyInfo(tokenId, newRecipient, newPercentage);
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                Metadata Functions
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IMutableMetadata
+    function updateTokenUri(uint256 tokenId, string calldata newUri) external {
+        if (!_exists(tokenId)) revert TokenDoesntExist();
+        if (msg.sender != owner() && !hasRole(ADMIN_ROLE, msg.sender) && !hasRole(APPROVED_MINT_CONTRACT, msg.sender)) {
```

Comment:
Wouldn't this revert if mint contract was not approved?

Reply:
Even if the contract owner was trying to update the metadata directly, and not through a mint contract

Reply:
if the mint contract is not approved, yes, it would revert. But if the contract owner calls, the `AND` statement will always be false and skipp the revert.